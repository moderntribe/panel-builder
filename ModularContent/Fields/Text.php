<?php


namespace ModularContent\Fields;

use ModularContent\Panel;

/**
 * Class Text
 *
 * @package ModularContent\Fields
 *
 * A basic text field.
 */
class Text extends Field {

	protected $input_width = 12;
	protected $layout      = 'full';


	public function __construct( array $args = [] ) {
		$this->check_input_width( $args );
		$this->check_layout( $args );

		$this->defaults['input_width'] = $this->input_width;
		$this->defaults['layout']      = $this->layout;

		parent::__construct( $args );
	}

	protected function check_layout( $args ) {
		if ( isset( $args['layout'] ) && $args['layout'] !== 'compact' && $args['layout'] !== 'full' && $args['layout'] !== 'inline' ) {
			throw new \LogicException( 'Layout argument can only be "compact" or "full".' );
		}
	}

	public function get_blueprint() {
		$blueprint                = parent::get_blueprint();
		$blueprint['input_width'] = (int) $this->input_width;
		$blueprint['layout']      = $this->layout;

		return $blueprint;
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * @param mixed $data
	 *
	 * @return string
	 */
	public function prepare_data_for_save( $data ) {
		return (string) parent::prepare_data_for_save( $data );
	}

	protected function check_input_width( $args ) {
		if ( isset( $args['input_width'] ) && ( isset( $args['layout'] ) && $args['layout'] === 'full' ) ) {
			throw new \LogicException( 'input_width only applies to compact layouts.' );
		}

		if ( isset( $args['input_width'] ) && ( $args['input_width'] < 1 || $args['input_width'] > 12 ) ) {
			throw new \LogicException( 'input_width argument must be between 1-12.' );
		}
	}
}