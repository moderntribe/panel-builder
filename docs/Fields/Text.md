## Text

A basic text field.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |
| `input_width` | `int` | The width of this input if part of a group. Can be 1-12 |
| `layout` | `string` | The layout for the field, either `compact`, `full`, or `inline` |

### Example 

```php
$field = new Text( [
    'label' => __('Name'),
    'name' => 'name',
    'description' => __( 'What are you called?' )
] );
```