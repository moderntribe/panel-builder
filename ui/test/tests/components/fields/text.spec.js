/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Text from 'components/fields/text';
import renderer from 'react-test-renderer';

describe('Text field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Text description="This is a test description" label="Some test label" />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Text />);
		expect(wrapper.type()).toEqual('div');
	});

	it('allows us to set props', () => {
		const wrapper = mount(<Text description="This is a test" />);
		expect(wrapper.props().description).toEqual('This is a test');
		wrapper.setProps({ description: 'Change props for test' });
		expect(wrapper.props().description).toEqual('Change props for test');
	});

	it('has a label', () => {
		const wrapper = shallow(<Text />);
		expect(wrapper.find('label').length).toEqual(1);
	});

	it('renders a text input', () => {
		const wrapper = shallow(<Text />);
		expect(wrapper.find('input[type="text"]').length).toEqual(1);
	});
});
