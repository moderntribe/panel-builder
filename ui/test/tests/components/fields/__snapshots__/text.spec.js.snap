exports[`Text field renders correctly 1`] = `
<div
  className="field panel-field">
  <label
    className="label panel-field-label">
    Some test label
    <span
      className="icon"
      onBlur={undefined}
      onClick={undefined}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined} />
  </label>
  <span
    className="inputContainer">
    <input
      name="modular-content-"
      onChange={[Function]}
      size="40"
      type="text"
      value="" />
  </span>
  <p
    className="description panel-field-description">
    This is a test description
  </p>
</div>
`;
